src/vocabulary/processor.rs

```diff
@@ -0,0 +1,402 @@
+/// to map back these type of characters into bytes, based on `CHAR_MAP`.
+///
+/// "ĠO" = [U+0120, U+004F] should be interpreted as [0x20, 0x4F] = " O"
+/// or
+/// "Ġal" = [U+0120, U+0061, U+006C] should be interpreted as [0x20, 0x61, 0x6C] = " al"
+///
+/// We'll use the following the mapping for this transition:
+/// ---
+/// 'Ā' == '\u{0100}' -> 0x00 == 0
+/// 'ā' == '\u{0101}' -> 0x01 == 1
+/// 'Ă' == '\u{0102}' -> 0x02 == 2
+/// ...
+/// 'Ğ' == '\u{011E}' -> 0x1E == 30
+/// 'ğ' == '\u{011F}' -> 0x1F == 31
+/// 'Ġ' == '\u{0120}' -> 0x20 == 32
+/// ---
+/// '!' == '\u{0021}' -> 0x21 == 33
+/// '"' == '\u{0022}' -> 0x22 == 34
+/// '#' == '\u{0023}' -> 0x23 == 35
+/// ...
+/// '|' == '\u{007C}' -> 0x7C == 124
+/// '}' == '\u{007D}' -> 0x7D == 125
+/// '~' == '\u{007E}' -> 0x7E == 126
+/// ---
+/// 'ġ' == '\u{0121}' -> 0x7F == 127
+/// 'Ģ' == '\u{0122}' -> 0x80 == 128
+/// 'ģ' == '\u{0123}' -> 0x81 == 129
+/// ...
+/// 'ŀ' == '\u{0140}' -> 0x9E == 158
+/// 'Ł' == '\u{0141}' -> 0x9F == 159
+/// 'ł' == '\u{0142}' -> 0xA0 == 160
+/// ---
+/// '¡' == '\u{00A1}' -> 0xA1 == 161
+/// '¢' == '\u{00A2}' -> 0xA2 == 162
+/// '£' == '\u{00A3}' -> 0xA3 == 163
+/// ...
+/// 'ª' == '\u{00AA}' -> 0xAA == 170
+/// '«' == '\u{00AB}' -> 0xAB == 171
+/// '¬' == '\u{00AC}' -> 0xAC == 172
+/// ---
+/// 'Ń' == '\u{0143}' -> 0xAD == 173
+/// ---
+/// '®' == '\u{00AE}' -> 0xAE == 174
+/// '¯' == '\u{00AF}' -> 0xAF == 175
+/// '°' == '\u{00B0}' -> 0xB0 == 176
+/// ...
+/// 'ý' == '\u{00FD}' -> 0xFD == 253
+/// 'þ' == '\u{00FE}' -> 0xFE == 254
+/// 'ÿ' == '\u{00FF}' -> 0xFF == 255
+/// ---
+static CHAR_MAP: Lazy<HashMap<char, u8>> = Lazy::new(|| {
+    let mut char_map = HashMap::with_capacity(256);
+    let mut key = 0x100u32;
+    for byte in 0..=255u8 {
+        let char = byte as char;
+        if matches!(
+            char, '!'..='~' | '\u{00A1}'..='\u{00AC}' | '\u{00AE}'..='\u{00FF}',
+        ) {
+            char_map.insert(char, byte);
+        } else if let Some(ch) = char::from_u32(key) {
+            char_map.insert(ch, byte);
+            key += 1;
+        }
+    }
+    char_map
+});
+
+/// Token processor to adjust tokens according to the tokenizer's level.
+#[derive(Debug)]
+pub(crate) struct TokenProcessor {
+    level: TokenProcessorLevel,
+}
+
+/// Recognizes different tokenizer's levels.
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum TokenProcessorLevel {
+    /// Matches byte level tokenizer (e.g., gpt2).
+    Byte,
+    /// Matches byte fallback tokenizer (e.g., llama), which have <0x__> tokens for
+    /// all __ >= 0x80 to represent incomplete UTF-8 sequences.
+    ByteFallback(Mods),
+}
+
+/// Modifications to be applied by `TokenProcessor`of `ByteFallback` level.
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct Mods {
+    spacechar: char,
+}
+
+/// Default string modification to be applied by `TokenProcessor` of `ByteFallback` level.
+static DEFAULT_MODS: Mods = Mods { spacechar: ' ' };
```

Comment:
This can be under `Mods` type like `Mods::default()` or `Mods::DEFAULT`.

Reply:
Well, associated `static` items are not allowed and they're static, which is better.

Reply:
Actually, thanks for this question, because I'm not even sure if `default()` would be less optimized than `const` before measuring it. But `static` for sure adds confusion here, since it's mostly used as a global variable in practice, while `default` on the other hand is exactly expected. I will adjust it!